python/ray/tune/function_runner.py

```diff
@@ -41,105 +47,184 @@ def __call__(self, **kwargs):
+            try:
+                # report the error but avoid indefinite blocking which would
+                # prevent the exception from being propagated in the unlikely
+                # case that something went terribly wrong
+                err_type, err_value, err_tb = sys.exc_info()
+                err_tb = err_tb.format_exc()
+                self._error_queue.put((err_type, err_value, err_tb),
+                                      block=True,
+                                      timeout=ERROR_REPORT_TIMEOUT)
+            except queue.Full:
+                logger.critical((
+                        "Runner Thread was unable to report error to main "
+                        "function runner thread. This means a previous error "
+                        "was not processed. This should never happen."))
             raise e
-        finally:
-            self._status_reporter._done = True
 
 
 class FunctionRunner(Trainable):
-    """Trainable that runs a user function returning training results.
+    """Trainable that runs a user function reporting results.
 
     This mode of execution does not support checkpoint/restore."""
 
     _name = "func"
-    _default_config = DEFAULT_CONFIG
 
     def _setup(self, config):
-        entrypoint = self._trainable_func()
-        self._status_reporter = StatusReporter()
-        scrubbed_config = config.copy()
-        for k in self._default_config:
-            if k in scrubbed_config:
-                del scrubbed_config[k]
-        self._runner = _RunnerThread(entrypoint, scrubbed_config,
-                                     self._status_reporter)
-        self._start_time = time.time()
-        self._last_reported_timestep = 0
-        self._runner.start()
+        # Semaphore for notifying the reporter to continue with the computation
+        # and to generate the next result.
+        self._continue_semaphore = threading.Semaphore(0)
+
+        # Queue for passing results between threads
+        self._results_queue = queue.Queue(1)
+
+        # Queue for passing errors back from the thread runner. The error queue
+        # has a max size of one to prevent stacking error and force error
+        # reporting to block until finished.
+        self._error_queue = queue.Queue(1)
+
+        self._status_reporter = StatusReporter(self._results_queue,
+                                               self._continue_semaphore)
+
+        config = config.copy()
+
+        def entrypoint():
+            return self._trainable_func(config, self._status_reporter)
+
+        # the runner thread is not started until the first call to _train
+        self._runner = _RunnerThread(entrypoint, self._error_queue)
 
     def _trainable_func(self):
         """Subclasses can override this to set the trainable func."""
 
         raise NotImplementedError
 
     def _train(self):
-        time.sleep(
-            self.config.get("script_min_iter_time_s",
```

Comment:
nit: `script_min_iter_time_s` is no longer a relevant parameter anywhere in the codebase right? Would it be possible to remove? Otherwise we can just make an issue and address a cleanup later.

Reply:
Do you mean remove it from the tests since it is not longer in the function runner? If so, I just clean the test a bit more and removed it.

Reply:
I've removed all of them found in test_trial_runner.py. If there are any others hiding around, I'm for the idea of creating a separate issue.